cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1382,6 +1385,76 @@ func maybeMountSave(disk disks.Disk, rootdir string, encrypted bool, mountOpts *
 	return true, nil
 }
 
+func generateMountsModeClassic(mst *initramfsMountsState, label string) error {
+	// Mount ESP
+	// Must be where the kernel came from, without fallbacks
+	espuuid, err := bootFindPartitionUUIDForBootedKernelDisk()
+	if err != nil {
+		return err
+	}
+	espSrc := filepath.Join("/dev/disk/by-partuuid", espuuid)
+	espOpts := &systemdMountOptions{
+		// vfat needs fsck to mount ESP that was unsafely
+		// unmounted
+		NeedsFsck: true,
+		// do not persist mount into running system, otherwise
+		// regular fstab cannot mount ESP at /boot/efi as
+		// expected
+		Ephemeral: true,
+	}
+	if err := doSystemdMount(espSrc, boot.InitramfsUbuntuSeedDir, espOpts); err != nil {
+		return err
+	}
+
+	// get the disk that we mounted the ESP from as a reference
+	// point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+	if err != nil {
+		return err
```

Comment:
Just a doubt, please ignore me if this doesn't apply: when we return because of an error, should we unwind the operations successfully performed so far? Is there a possibility that this method fails, then gets executed again (after fixing the cause that made it fail), but still fails because of the previous operations taht we didn't clean up?

Reply:
Note that none of the other modes unwind things. By default we reboot if we fail here. Emergency shell & debug shells are not available (requires "dangerous" on the kernel cmdline). Thus there is actually no way to "fixing the cause that made it fail" without a reboot.

Reply:
Makes sense, thanks for the explanation!